src/schema/schema_drop.c

```diff
@@ -174,31 +174,83 @@ __drop_table(WT_SESSION_IMPL *session, const char *uri, const char *cfg[])
  *     Drop a tiered store.
  */
 static int
-__drop_tiered(WT_SESSION_IMPL *session, const char *uri, const char *cfg[])
+__drop_tiered(WT_SESSION_IMPL *session, const char *uri, bool force, const char *cfg[])
 {
+    WT_CONFIG_ITEM cval;
     WT_DATA_HANDLE *tier;
     WT_DECL_RET;
     WT_TIERED *tiered;
     u_int i;
+    const char *name;
+    bool remove_files;
+
+    WT_RET(__wt_config_gets(session, cfg, "remove_files", &cval));
+    remove_files = cval.val != 0;
 
+    name = NULL;
     /* Get the tiered data handle. */
     WT_RET(__wt_session_get_dhandle(session, uri, NULL, NULL, WT_DHANDLE_EXCLUSIVE));
     tiered = (WT_TIERED *)session->dhandle;
 
-    /* Drop the tiers. */
-    for (i = 0; i < WT_TIERED_MAX_TIERS; i++) {
-        tier = tiered->tiers[i].tier;
-        if (tier != NULL)
-            WT_ERR(__wt_schema_drop(session, tier->name, cfg));
+    /*
+     * We cannot remove the objects on shared storage as other systems may be accessing them too.
+     * Remove the current local file object, the tiered entry and all bucket objects from the
+     * metadata only.
+     */
+    tier = tiered->tiers[WT_TIERED_INDEX_LOCAL].tier;
+    if (tier != NULL) {
+        __wt_verbose(session, WT_VERB_TIERED, "DROP_TIERED: drop local object %s", tier->name);
+        WT_WITHOUT_DHANDLE(session,
+          WT_WITH_HANDLE_LIST_WRITE_LOCK(
+            session, ret = __wt_conn_dhandle_close_all(session, tier->name, true, force)));
+        WT_ERR(__wt_metadata_remove(session, tier->name));
+        if (remove_files)
+            WT_TRET(__wt_meta_track_drop(session, tier->name));
     }
 
-    ret = __wt_metadata_remove(session, uri);
+    /* Close any dhandle and remove any tier: entry from metadata. */
+    tier = tiered->tiers[WT_TIERED_INDEX_SHARED].tier;
+    if (tier != NULL) {
+        __wt_verbose(session, WT_VERB_TIERED, "DROP_TIERED: drop shared object %s", tier->name);
+        WT_WITHOUT_DHANDLE(session,
+          WT_WITH_HANDLE_LIST_WRITE_LOCK(
+            session, ret = __wt_conn_dhandle_close_all(session, tier->name, true, force)));
+        WT_ERR(__wt_metadata_remove(session, tier->name));
+    }
 
-err:
-    F_SET(session->dhandle, WT_DHANDLE_DISCARD);
+    /*
+     * We remove all metadata entries for both the file and object versions of an object. The local
+     * retention means we can have both versions in the metadata. Ignore WT_NOTFOUND.
+     *
+     * Object ids start at 1.
+     */
+    for (i = 1; i < tiered->current_id; ++i) {
+        WT_ERR(__wt_tiered_name(session, &tiered->iface, i, WT_TIERED_NAME_LOCAL, &name));
+        __wt_verbose(session, WT_VERB_TIERED, "DROP_TIERED: remove object %s from metadata", name);
+        WT_ERR_NOTFOUND_OK(__wt_metadata_remove(session, name), false);
```

Comment:
Maybe also call `__wt_meta_track_drop`?

Reply:
I had to think about this, but I think you're right, we need that here too. Fixed.